### DISCUSSION TITLE
Run tests tagged with "smoke-tests" first?

### DISCUSSION THREAD

**Discussion by pandoras-toolbox** [junit-team/junit-framework#3683] [General]
Is it possible to implement a JUnit extension which runs tests which are tagged with "smoke-tests" before all other tests (in a single run)?

Or how can it be done _**without**_ running first the tests tagged with "smoke-test" and after that running all other tests?

**Comment by sbrannen** [comment]
Yes, that's possible.

You need to implement a `ClassOrderer` that retrieves the `@Tag` annotations via the  `ClassDescriptor` and sort the results however you wish.

You might also find this blog useful: https://www.wimdeblauwe.com/blog/2021/02/12/junit-5-test-class-orderer-for-spring-boot/

**Reply by pandoras-toolbox** [reply]
@sbrannen That seems only to work as expected if tests are not executed in parallel, right? At least that is how it behaves when I implemented that for myself. We run tests fully in parallel and not sequentially. I hoped that even then it would consider the desired order.

**Reply by pandoras-toolbox** [reply]
It turns out I had to disable my custom `ClassOrderer` and only use my custom `MethodOrderer` with concurrent/concurrent parallelism to achieve what I wanted.

**Comment by pandoras-toolbox** [comment]
Thank you, I will try that out.

It is because currently our E2E smoke tests cannot be run in parallel when the other tests are running. A quick fix would be to run the smoke test in a separate pipeline stage, but then we would lose having a single test result report, for which we use Allure.

Running tests in order of priority (there is an Allure annotation for that) can be useful, not just for our current issue.

**Comment by sbrannen** [comment]
Closing in light of https://github.com/junit-team/junit5/discussions/3683#discussioncomment-8494399.
